### PR TITLE
hand-coded ROA example for the cubic polynomial

### PR DESCRIPTION
**Pull request by RussTedrake**
@soonho-tri , @FischerGundlach, @frankpermenter  -- this example is typed in, but runs (for all intents and purposes) forever.

The problem is calling `NewSosPolynomial` with `degree=8` (which is a perfectly small degree, especially since i only have one indeterminate!).   The program effectively hangs on the line
```
  const symbolic::Polynomial p{x.dot(Q_poly * x)};  // p = xᵀ * Q_poly * x.
```
(`Q_poly` ends up being 210x210).  

I figured i'd share this as a PR instead of an issue so you can run/comment on the code if you like.   

Is this going to get magically much better with @FischerGundlach 's work-in-progress?  Should we expect to hit the performance bottleneck so soon because of the naive implementation (without even the newton polytope)?

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6709)

### CODE REVIEW

**Comment by soonho-tri** [comment]
Hi Russ, thanks for the report. I'm investigating the issue. I'll be back with more information.

**Comment by soonho-tri** [comment]
We call [symbolic::Polynomial::CheckInvariant()](https://github.com/RobotLocomotion/drake/blob/master/drake/common/symbolic_polynomial.cc#L427-L438) frequently in the codebase while it turns out to be pretty expensive. Turning it off for release builds reduce the running time of the example to a minute. I'll find a way to make it faster.

BTW, /cc @hongkai-dai.

**Comment by soonho-tri** [comment]
@RussTedrake , I have a version which only takes 0.4 sec to have the following result (without disabling the `CheckInvariant()` that I mentioned above):
```
Verified that 1*x(0)^2 < 1 is in the region of attraction.
```

The code is available at https://github.com/soonho-tri/drake/blob/pr-cubic_polynomial/drake/examples/cubic_polynomial/region_of_attraction.cc

Notes
------

1. We have `MathematicalProgram::NewFreePolynomial()` function, which can reduce the following lines 

    ```c++
    const VectorX<symbolic::Monomial> m(symbolic::MonomialBasis({x}, 4));
    const auto lambda =
        prog.NewContinuousVariables<5>("c").cast<symbolic::Monomial>().dot(
            m.cast<symbolic::Polynomial>());
    ```
    into
    ```c++
    const Polynomial lambda{prog.NewFreePolynomial({x}, 4, "c")};
    ```
    Note that, `lambda = prog.NewFreePolynomial()` adds the constraint that `lambda` is SOS internally. As a result, we don't need to call `prog.AddSosConstraint(lambda)` separately.

2. `prog.NewContinuousVariables<4>("c").cast<symbolic::Monomial>()` returns `[c₀, c₁, c₂, c₃, c₄]`. However, each `cᵢ` is treated as an indeterminate instead of a decision variable. Here, the fundamental issue is that a `symbolic::Variable` does not have information if a variable is a decision variable or an indeterminate. This is the reason why we do not allow a multiplication between a polynomial and a variable (and also a monomial and a variable).

3. By default, `symbolic::Polynomial(const Expression& e)` constructs a polynomial from `e` *assuming* that *all* variables in `e` are *indeterminates*. To specify a set of indeterminates, we provide another constructor `symbolic::Polynomial(const Expression& e, const Variables& indeterminates)`. For example, in this PR, we have `symbolic::Polynomial(rho)` which should be `symbolic::Polynomial(rho, {})`.

TODOs / Proposals
-------------------

1. If we agree that polynomial plays an important role in Drake, we can add a 'kind' field in `symbolic::Variable` to distinguish a decision variable from an indeterminate intensionally. This will solve the problems we had in this example. A downside is that a `symbolic::Variable` becomes slightly heavier.

2. For now, we have the hierarchy in the `symbolic` namespace.

    screen shot 2017-07-31 at 2 56 55 am

    Note that `Monomial` and `Polynomial` are not derived by common operations, and not included in the diagram. When we add the `kind` information in `symbolic::Variable`, we can bring `symbolic::Monomial` and `symbolic::Polynomial` into the hierarchy as follows: 

    screen shot 2017-07-31 at 2 57 25 am

    Note that If we move to this direction, we can think about promoting `symbolic::Polynomial` as another `ScalarType` of Drake.

**Comment by hongkai-dai** [comment]
That is a very clear document. Thanks @soonho-tri !

I just want to clarify one thing. From `Monomial` to `Polynomial` class, it can take more operations than `+, - *`. The reason is that the coefficients of `Polynomial` are not just doubles, but can be expressions. So `sin(a) * x² + eᵇ * x * y` is a valid polynomial, with indeterminate `x` and `y`. This polynomial takes operation like `sin` and `exp`. This also means that `log, pow, exp` does not always map `Variable` to `Expression`, but can also map to `Polynomial`.

**Comment by RussTedrake** [comment]
Re: cleaned up cubic poly example:  woot!  @soonho-tri, your version looks great.  feel free to PR it to my branch or directly to master. (i think it's ready; sorry I handn't done clang-format, etc on it yet).  One minor comment:  can update the [other lambda comment ](https://github.com/soonho-tri/drake/blob/pr-cubic_polynomial/drake/examples/cubic_polynomial/region_of_attraction.cc#L69) to the special characters, too?

**Comment by RussTedrake** [comment]
Re: NewFreePolynomial.  Awesome that the method exists, that's much better.  But i think it should definitely not add the lambda is SOS constraint automatically, given that method name?  And there are plenty of use cases where we want the polynomial to be sign indefinite (e.g. for s-procedure with equality constraints, instead of the inequality constraints in this example).  

I actually don't love the name "NewFreePolynomial", either.  That's a carry-over from SPOTLESS, and SPOT before that, which distinguished "free variables" from "indeterminates".  We call "free variables" "decision variables" in drake.  Since "NewPolynomialDecisionVariables" isn't right, I'd say maybe just "NewPolynomial" would be better?

**Comment by RussTedrake** [comment]
Re: Monomial/Polynomial vs the other symbolic classes.  I haven't thought as deeply about it as you, @soonho-tri , but as a user I was expecting the hierarchy to behave more like the updated hierarchy you've proposed.  It was confusing and frustrating to not be able to work very easily with monomials and polynomials.  I think that having whichever diagram we choose land in the documentation would help a lot, but given the naming I think new users (who don't read the docs first) will assume the second.  if it's viable, then I think it's probably the right way to go.

The other confusion in the current code is that there are two polynomial classes.  and symbolic::Expression has a method `ToPolynomial` that kicks out to the other polynomial class, not symbolic::Polynomial.  What's the roadmap for that going forward?

**Comment by soonho-tri** [comment]
Hi @RussTedrake,

Re: Cleaned up cubic poly example: I'll update my branch and open a PR to `master` this morning.

Re: `NewFreePolynomial`: Thanks for the comments. I will have a discussion with @hongkai-dai and @FischerGundlach, and open another issue with our plan. It should be aligned with your comments.

Re: Monomial/Polynomial vs the other symbolic classes: I will start implementing my proposal. I agree the current design is not easy to use.

Re: `drake::Polynomial<T>` vs `drake::symbolic::Polynomial`: I have a plan to unify them. I'll open another issue on the topic. In short, there will be a template class `drake::symbolic::Polynomial<typename CoeffType>`. You can think of the current `symbolic::Polynomial` as an instantiation of the new `symbolic::Polynomial<symbolic::Expression>`. The goal is to keep the functionalities of the current `symbolic::Polynomial` and the performance of `drake::Polynomial<T>`.

**Comment by soonho-tri** [comment]
> I just want to clarify one thing. From Monomial to Polynomial class, it can take more operations than +, - *. The reason is that the coefficients of Polynomial are not just doubles, but can be expressions. So sin(a) * x² + eᵇ * x * y is a valid polynomial, with indeterminate x and y. This polynomial takes operation like sin and exp. This also means that log, pow, exp does not always map Variable to Expression, but can also map to Polynomial.

@hongkai-dai , you're right. I'll open another issue to explain the details of the proposal.

**Comment by RussTedrake** [comment]
excellent!  thanks @soonho-tri  (fwiw, I'm on slack if it helps to chat)

Re: Decision variable vs indeterminate confusion.  I'm still confused about that.  Both decision variables and indeterminates are simply aliases of symbolic::Variable, i believe?  A monomial, polynomial, expression, and/or formula shouldn't care (they are just variables here)?  Only the solver holds the distinction, so when you pass a symbolic expression in that is polynomial in some variables, it knows how to handle them as either decision variables (must have dependency deg <= 1) or indeterminates (arbitrary degree).  Can you help me clean up this thinking if it's wrong?

In other words, I don't understand why `Polynomial(rho, {})` is needed?

**Comment by soonho-tri** [comment]
@RussTedrake : Let's keep the discussion here in Github, 1) for the record, and 2) to keep it open for other people (i.e. @hongkai-dai).

> Both decision variables and indeterminates are simply aliases of symbolic::Variable, i believe? 

Your observation is correct. There is no distinction between decision variables and indeterminates in symbolic `Variable`, `Expression`, and `Formula`.

> A monomial, polynomial, expression, and/or formula shouldn't care (they are just variables here)?

In symbolic `Monomial` and `Polynomial`, the notion of variable kind (decision variables vs indeterminates) is embedded. Let me first illustrate their internal structures, and give you an application where this distinction is useful.

1. A monomial is a map from an indeterminate to its integer power. Note that it does not include the coefficient part. As a result, every variable appears in a monomial is an indeterminate. For example, `x`, `y`, `x²y` are Monomials assuming `x` and `y` being indeterminates. In contrast, `a`, `b`, `a²b` are Polynomials (`a` = `a * 1`, `b` = `b * 1`, `a²b = a²b * 1`), not Monomials, given `a` and `b` are decision variables. Note that `1` is a Monomial (an empty Monomial).

2. A polynomial `p` is a sum of products, `p = ∑ᵢ cᵢ * mᵢ` where `cᵢ` is a coefficient and `mᵢ` is a monomial. We have an invariant: `cᵢ` only includes decision variables while `mᵢ` only includes indeterminates. That is, `∀ i. vars(cᵢ) ∩ vars(mᵢ) = ∅` holds.

3. This separation helps us do the process of coefficients matching efficiently. Please consider the following example while using the same assumption - `x` is an indeterminate while `a`, `b`, `c` are decision variables:  
    ```
    p₁ = ax² + bx + c
    p₂ = cx² + ax + d
    ```
    With the embedded information, we can compute that `p₁ == p₂` is `(a = c) ∧ (b = a) ∧ (c = d)`, without relying on extra information.

4. Note that I've been thinking about having `symbolic::Polynomial<T>` where `T` is a type of  coefficients (we will have at least two instantiations `T = double` and `T = Expression`). In this design, the separation between coefficients and monomials in the internal representation is essential.

I might be wrong though. It could be the case that we can simply postpone the distinction of decision variables and indeterminates until we have a request for the coefficients matching where we supply extra information from a MathematicalProgram. I haven't thought about the problem in that direction. I'll think about this direction more carefully.

@hongkai-dai , @FischerGundlach : As always, your comments are welcome.

**Comment by FischerGundlach** [comment]
> Re: NewFreePolynomial. Awesome that the method exists, that's much better. But i think it should definitely not add the lambda is SOS constraint automatically, given that method name? And there are plenty of use cases where we want the polynomial to be sign indefinite (e.g. for s-procedure with equality constraints, instead of the inequality constraints in this example).

One misconception here to clear up. In deed, the function `NewFreePolynomial` does not imply a `SOS` constraint on the decision variables but returns as you, @RussTedrake , expected a polynomial similar to the `freePolynomial` from SPOTLESS, i.e. a polynomial of the form `c_0 + c_1*x_1 + c_2*x_2...` where `c_i` are decision variables and `x_i` are indeterminate. 

I guess, that @hongkai-dai meant `NewSosPolynomial` which does bind the decision variables with a SOS constraint, i.e. internally we find the familiar form of `m^T*Q*m`, where `m` is a monomial vector `drake::VectorX<symbolic::monomials>` and `Q` a symmetric decision matrix with `AddPositiveSemidefiniteConstraint(Q)`.

**Comment by soonho-tri** [comment]
Oops, it was my mistake. Yes, @FischerGundlach is correct. I meant `NewSosPolynomial` and the code should be:
```c++
  const Polynomial lambda{prog.NewSosPolynomial({x}, 4).first};
```
and it does increase the running time to around 1.4sec.

**Comment by soonho-tri** [comment]
Closing this with https://github.com/RobotLocomotion/drake/pull/6715. Note that I measure the running time in my office and its around `0.3` sec. I was mobile when I got `1.4` sec and I blame MOSEK license checking.

**Comment by RussTedrake** [comment]
thank you all!
